### PR TITLE
Feature/automatedlogin

### PR DESCRIPTION
**Pull request by biscon**
- implemented support for Automated login, config file parsing and user selection
- refactored library auth engines into AuthCodeEngine and AutomatedAuthEngine
- rewrote the way the library deals with activity launchers
- implemented getting the host application context via content provider
- removed no longer needed arguments to Auth Engine configuration builders
- updated documentation to reflect the new usage

### CODE REVIEW

File: ticketauth/src/main/kotlin/dk/ufst/ticketauth/TicketAuthInitProvider.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import android.content.ContentProvider
+import android.content.ContentValues
+import android.content.Context
+import android.database.Cursor
+import android.net.Uri
+import android.util.Log
+
+internal var hostApplicationContext: Context? = null
+
+class TicketAuthInitProvider: ContentProvider() {
```
Comment (patrickusiewicz) [inline ticketauth/src/main/kotlin/dk/ufst/ticketauth/TicketAuthInitProvider.kt:12]: Because we already have a `TicketAuth.setup`, I wouldn't mind to pass a `ApplicationContext` into that setup call. Less code in this library and one extra parameter for the client to provide.
But this also works.

File: ticketauth/build.gradle.kts [hunk 2]
```diff
@@ -9,10 +11,15 @@ android {
...
         minSdk = 26
-        targetSdk = 31
+        targetSdk = 32
 
         testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
         consumerProguardFiles("consumer-rules.pro")
+        buildConfigField(
+            "String",
+            "TICKETAUTH_VERSION",
+            "\"1.1.0\""
```
Comment (patrickusiewicz) [inline ticketauth/build.gradle.kts:21]: We might need to add a release section in the `README` that specifies one need to update this before a release.

File: ticketauth/build.gradle.kts [hunk 3]
```diff
@@ -1,3 +1,5 @@
+@file:Suppress("UnstableApiUsage")
```
Comment (patrickusiewicz) [inline ticketauth/build.gradle.kts:1]: Why is this needed?

File: settings.gradle.kts [hunk 4]
```diff
@@ -1,3 +1,4 @@
+@file:Suppress("UnstableApiUsage")
```
Comment (patrickusiewicz) [inline settings.gradle.kts:1]: On my computer it looks like this is not needed.

File: docs/README.md [hunk 5]
```diff
@@ -64,6 +72,32 @@ If you specify a custom redirectUri you need to define a manifest override as we
...
 file.__
 
+### Automated Auth
+Setup the library for Automated login in which the library reads and present users from
+a json configuration file. This file contains the necessary data to request an access token
+from a third party endpoint. This feature is meant to ease testing.
+
+___Automated login doesn't support token refresh and will default to relogin whenever the token expires.___
+
+```
```
Comment (patrickusiewicz) [inline docs/README.md:82]: You can add kotlin as the language in the code block (see [here](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting)).

File: ticketauth/src/main/kotlin/dk/ufst/ticketauth/automated/SelectUserActivity.kt [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    @SuppressLint("SetTextI18n")
+    private fun buildContentView(users: JSONArray): View {
+        val bgColor = Color.parseColor("#14143C")
+        val textColor = Color.parseColor("#FFFFFF")
+        val ll = LinearLayout(this)
+        ll.orientation = LinearLayout.VERTICAL
+        ll.layoutParams = LayoutParams(MATCH_PARENT, MATCH_PARENT)
+        ll.setPadding(0, toDp(16f).toInt(), 0, 0)
+        ll.setBackgroundColor(bgColor)
+        TextView(this)
```
Comment (patrickusiewicz) [inline ticketauth/src/main/kotlin/dk/ufst/ticketauth/automated/SelectUserActivity.kt:43]: Cool that you have kept it simple here and not used Compose.

File: docs/README.md [hunk 7]
```diff
@@ -156,6 +190,8 @@ If you want to run the oauth logout flow call:
 authenticator.logout()
 ```
 
+___When using Automated Auth no actual login endpoint is called. Instead local auth state is invalidated.___
```
Comment (patrickusiewicz) [inline docs/README.md:193]: Isn't there some endpoint being called? It looks like `MicroHttpCLient` is used as part of Automated Auth to get a token from tokenUrl?

**Review by patrickusiewicz — APPROVED** [review]
The automated auth is an nice feature, thank you for that!

I added some comments, but nothing blocking.

File: ticketauth/build.gradle.kts (hunk 3 above)
Comment (biscon) [reply ticketauth/build.gradle.kts:1]: apparently if you use agp 7.x.x with latest version of Android Studio a lot of the regular functions like minSdk etc are marked @incubating, which they are gonna be for half a year or something, so I suppressed them for now. Let me know if you know a better solution or if I got @incubating wrong :)

File: settings.gradle.kts (hunk 4 above)
Comment (biscon) [reply settings.gradle.kts:1]: update to the newest version of android studio and get a lot of warnings about stuff being marked @incubating see above comment

File: ticketauth/src/main/kotlin/dk/ufst/ticketauth/automated/SelectUserActivity.kt (hunk 6 above)
Comment (biscon) [reply ticketauth/src/main/kotlin/dk/ufst/ticketauth/automated/SelectUserActivity.kt:43]: Thanks but it dawned upon that I AM including appcompat (for activity result launchers) and here I use the regular framework. I don't see any problem with this simple usage, but might be nicer to use appcompat in a future update. .im gonna make stuff like colors configurable anyway as you requested in the demo :)

File: docs/README.md (hunk 7 above)
Comment (biscon) [reply docs/README.md:193]: my bad, this is for logout, but I wrote login endpoint in the note.. ill correct it to logout endpoint.
The micro client is used as you describe, url is in the config file

File: ticketauth/build.gradle.kts (hunk 2 above)
Comment (biscon) [reply ticketauth/build.gradle.kts:21]: good input, will do :+1:

File: ticketauth/src/main/kotlin/dk/ufst/ticketauth/TicketAuthInitProvider.kt (hunk 1 above)
Comment (biscon) [reply ticketauth/src/main/kotlin/dk/ufst/ticketauth/TicketAuthInitProvider.kt:12]: I agree I may have went a little overboard with that feature given that as you mentioned, we don't explicitly need to setup some stuff before the rest of the app is initialized :).

File: docs/README.md (hunk 5 above)
Comment (biscon) [reply docs/README.md:82]: nice, will do :+1:
